### DISCUSSION TITLE
Custom force-stop during uiautomator2 startup

### DISCUSSION THREAD

**Discussion by v-mwalk** [appium/appium#16555] [Q&A] [answered]
Hi all,
I have an issue whereby if a test has left chrome hanging (IE. AUT had opened chrome and left it open) then Appium2/uiautomator2 cannot start the native app during test startup of subsequent test:-

```
[35m[UiAutomator2][39m Starting 'com.mynativeapp.mobile.develop/com.mynativeapp.mobile.dev.ui.DevOptionsActivity and waiting for 'com.mynativeapp.mobile.develop/com.mynativeapp.mobile.dev.ui.DevOptionsActivity'
[debug] [35m[ADB][39m Running '/Users/mwalk/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell am start -W -n com.mynativeapp.mobile.develop/com.mynativeapp.mobile.dev.ui.DevOptionsActivity -S'
[debug] [35m[ADB][39m Waiting up to 20000ms for activity matching pkg: 'com.mynativeapp.mobile.develop' and activity: 'com.mynativeapp.mobile.dev.ui.DevOptionsActivity' to be focused
[debug] [35m[ADB][39m Possible activities, to be checked: 'com.mynativeapp.mobile.dev.ui.DevOptionsActivity', 'com.mynativeapp.mobile.develop.com.mynativeapp.mobile.dev.ui.DevOptionsActivity'
[debug] [35m[ADB][39m Getting focused package and activity
[debug] [35m[ADB][39m Running '/Users/myuser/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell dumpsys window displays'
[debug] [35m[ADB][39m Found package: 'com.android.chrome' and fully qualified activity name : 'org.chromium.chrome.browser.customtabs.CustomTabActivity'
[debug] [35m[ADB][39m Incorrect package and activity. Retrying.
```

I cannot fix this within Appium as the earliest access I have is in the @before hook.  Which is never reached as the issue is happening as uiautomator2 is trying the launch the app as part of the driver startup.

So, is there a way I can - using capabilities/settings - say to Appium/UIAutomator2 do a force-stop on pkg 'xyzzy' (in my case, it would be 'com.android.chrome' OR run <this> defined mobile shell command during startup?

I currently have a workaround, whereby I'm running a nodejs execute `adb -s emulator-5554 shell am force-stop 'com.android.chrome'` outside webdriverio/appium.  However, not sure this will work in a pipeline with multiple agents...

**Comment by KazuCocoa** [comment, marked as answer]
You can start a session without `app` and `bundleid`. Then, appium starts a session without launching them.
After that, you can install an app/launch a package via http://appium.io/docs/en/commands/device/app/install-app/, https://appium.io/docs/en/commands/device/app/activate-app/ and https://github.com/appium/appium-uiautomator2-driver#mobile-startactivity etc

They could help what you'd like to do.
